File: LeetCodeRecord/MethodFirst/94-InorderTraversal.java
package MethodFirst;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

class InorderTraversal {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) {
            val = x;
        }
    }
    
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        inorderRecur(root, res);
        
        return res;
    }

    public void inorderRecur(TreeNode root, List<Integer> res) {
        if (root == null) {
            return;
        }

        inorderRecur(root.left, res);

        res.add(root.val);

        inorderRecur(root.right, res);
    }

    public List<Integer> inorderTraversal2(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode curr = root;
        
        while (curr != null || !stack.isEmpty()) {
            if (curr != null) {
                // stack 中存的是本身和右子树没有访问过的
                stack.push(curr);
                curr = curr.left;
            } else {
                TreeNode temp = stack.pop();
                res.add(temp.val);
                curr = temp.right;
            } 
        }

        return res;
    }

    public List<Integer> inorderTraversal3(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        TreeNode curr = root;

        while (curr != null) {
            if (curr.left != null) {
                TreeNode pre = curr.left;

                while (pre.right != null && pre.right != curr) {
                    pre = pre.right;
                }

                if (pre.right != null) {
                    pre.right = null;
                    res.add(curr.val);
                    curr = curr.right;
                } else {
                    pre.right = curr;
                    curr = curr.left;
                }
            } else {
                res.add(curr.val);
                curr = curr.right;
            }
        }

        return res;
    }
}

File: Top200/94-InorderTraversal.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type Stack struct {
	arr  []*TreeNode
	size int
}

func NewStack() *Stack {
	return &Stack{
		arr:  []*TreeNode{},
		size: 0,
	}
}

func (s *Stack) Pop() *TreeNode {
	if s.size <= 0 {
		return nil
	}

	node := s.arr[s.size-1]
	s.arr = s.arr[:s.size-1]
	s.size--

	return node
}

func (s *Stack) Top() *TreeNode {
	if s.size <= 0 {
		return nil
	}

	return s.arr[s.size-1]
}

func (s *Stack) Push(node *TreeNode) {
	s.arr = append(s.arr, node)
	s.size++
}

func (s *Stack) IsEmpty() bool {
	if s.size <= 0 {
		return true
	}

	return false
}

func inorderTraversal(root *TreeNode) []int {
	res := []int{}

	inorder(root, &res)

	return res
}

func inorder(root *TreeNode, res *[]int) {
	if root == nil {
		return
	}

	inorder(root.Left, res)

	*res = append(*res, root.Val)

	inorder(root.Right, res)
}

func inorderTraversal2(root *TreeNode) []int {
	res := []int{}
	stack := NewStack()

	if root == nil {
		return res
	}

	stack.Push(root)

	for !stack.IsEmpty() {
		node := stack.Top()

		if node.Left != nil {
			stack.Push(node.Left)
			node.Left = nil
			continue
		}

		res = append(res, node.Val)
		stack.Pop()

		if node.Right != nil {
			stack.Push(node.Right)
		}
	}

	return res
}
